packages/keyring-eth-qr/src/device.ts

```diff
@@ -0,0 +1,613 @@
+  /**
+   * The derivation paths for each hex-encoded address in the device
+   */
+  paths: Record<Hex, string>;
+};
+
+/**
+ * An address with its corresponding index.
+ */
+export type IndexedAddress = {
+  address: Hex;
+  index: number;
+};
+
+/**
+ * The details of the source CryptoAccount or CryptoHDKey
+ * that the Device uses to derive accounts.
+ */
+export type DeviceDetails = CommonDeviceDetails &
+  (HDModeDeviceDetails | AccountModeDeviceDetails);
+
+/**
+ * Get the fingerprint of the source CryptoAccount or CryptoHDKey
+ *
+ * @param source - The source CryptoAccount or CryptoHDKey
+ * @returns The fingerprint of the source
+ */
+function getFingerprintFromSource(source: CryptoAccount | CryptoHDKey): string {
+  return source instanceof CryptoAccount
+    ? source.getMasterFingerprint()?.toString('hex')
+    : source.getParentFingerprint()?.toString('hex');
+}
+
+/**
+ * Get fingerprint, account paths and names from the a CryptoAccount
+ *
+ * Note: This function emulates the behavior of the `@keystonehq/base-eth-keyring`
+ * library when dealing with CryptoAccount objects (for backwards compatibility
+ * reasons). Though, the way it retrieves `name` and `keyringAccount` is questionable,
+ * as `name` and `keyringAccount` are updated after each descriptor discovery, effectively
+ * returning the last descriptor's `name` and `keyringAccount`.
+ *
+ * @param source - The source CryptoAccount
+ * @returns The paths
+ */
+function readCryptoAccountOutputDescriptors(source: CryptoAccount): {
+  xfp: string;
+  paths: Record<Hex, string>;
+  name: string;
+  keyringAccount: string;
+} {
+  const descriptors = source.getOutputDescriptors();
+
+  if (!descriptors || descriptors.length === 0) {
+    throw new Error('No output descriptors found in CryptoAccount');
+  }
+
+  let name = '';
+  let keyringAccount = '';
+  const paths = descriptors.reduce(
+    (descriptorsPaths: Record<Hex, string>, current) => {
+      const hdKey = current.getHDKey();
+      if (hdKey) {
+        const path = `M/${hdKey.getOrigin().getPath()}`;
+        const address = getChecksumAddress(
+          add0x(
+            Buffer.from(publicToAddress(hdKey.getKey(), true)).toString('hex'),
+          ),
+        );
+        descriptorsPaths[address] = path;
+        name = hdKey.getName();
+        keyringAccount = hdKey.getNote();
+      }
+      return descriptorsPaths;
+    },
+    {},
+  );
+
+  return {
+    paths,
+    name,
+    keyringAccount,
+    xfp: getFingerprintFromSource(source),
+  };
+}
+
+export class Device {
+  readonly #requestScan: QrKeyringBridge['requestScan'];
+
+  #pairedDevice?: DeviceDetails | undefined;
+
+  constructor(requestScan: QrKeyringBridge['requestScan']) {
+    this.#requestScan = requestScan;
+  }
+
+  /**
+   * Initialize the Device with a source.
+   *
+   * @param source - The account source, which can be a CryptoAccount, CryptoHDKey, or a UR string.
+   */
+  init(source: DeviceDetails | string | SerializedUR): void {
```

Comment:
What if `init` is called more than once? Should we check to see if the Device has already been initialized and do nothing or throw an error in that case? (I don't know how we usually handle this type of thing elsewhere.)

Alternatively, does it make sense to move the initialization to the constructor?

Reply:
Nice, looks a lot better!